### PR TITLE
Support generics in docblock

### PR DESCRIPTION
**Pull request by tm1000**
This supports the following:

```
    /**
	 * @param Collection<int, TelephoneNumber> $telephoneNumbers
	 */
	public function myFunction(
		Collection $telephoneNumbers,
```

Where-as before it would say that the typehint needs to be `Collection<int, TelephoneNumber>` now it will correctly deduce that the type hint needs to be `Collection` and will not return an error

I based this off of https://github.com/squizlabs/PHP_CodeSniffer/pull/708 which is similar in scope. 

~~Unfortunately I was unable to run the unit tests because I'm on PHP 8.1 and something in the code continues to try to use the function `each` which does not exist in php 8.0~~

Edit: There is now a test!

### CODE REVIEW

**Comment by jrfnl** [comment]
> Unfortunately I was unable to run the unit tests because I'm on PHP 8.1 and something in the code continues to try to use the function `each` which does not exist in php 8.0

Please add tests for this feature.

To run the tests locally on PHP 8.1:
* Run `composer install --ignore-platform-reqs` to get PHPUnit 7.5.
* Then run `vendor/bin/phpunit tests/AllTests.php --no-configuration --bootstrap=tests/bootstrap.php --dont-report-useless-tests`

See: https://github.com/squizlabs/PHP_CodeSniffer/blob/3e6ce10d7f9092d0e7b9cd0a959aba3529952a27/.github/workflows/test.yml#L121-L142

**Comment by tm1000** [comment]
@jrfnl I already did that it does not work because of this library's usage of "each"

```shell
PHP Fatal error:  Uncaught Error: Call to undefined function each() in /usr/src/PHP_CodeSniffer/vendor/phpunit/phpunit/src/Util/Getopt.php:38
Stack trace:
#0 /usr/src/PHP_CodeSniffer/vendor/phpunit/phpunit/src/TextUI/Command.php(245): PHPUnit_Util_Getopt::getopt()
#1 /usr/src/PHP_CodeSniffer/vendor/phpunit/phpunit/src/TextUI/Command.php(120): PHPUnit_TextUI_Command->handleArguments()
#2 /usr/src/PHP_CodeSniffer/vendor/phpunit/phpunit/src/TextUI/Command.php(109): PHPUnit_TextUI_Command->run()
#3 /usr/src/PHP_CodeSniffer/vendor/phpunit/phpunit/phpunit(47): PHPUnit_TextUI_Command::main()
#4 {main}
  thrown in /usr/src/PHP_CodeSniffer/vendor/phpunit/phpunit/src/Util/Getopt.php on line 38
```

**Comment by tm1000** [comment]
@jrfnl got it to work! I had to delete the composer.lock file and vendor dir then run with `--ignore-platform-reqs`

Thanks for pointing that out. much appreciated. There is now a test!

File: src/Standards/Squiz/Sniffs/Commenting/FunctionCommentSniff.php [hunk 1]
```diff
@@ -419,7 +419,9 @@ protected function processParams(File $phpcsFile, $stackPtr, $commentStart)
 
                 // Check type hint for array and custom type.
                 $suggestedTypeHint = '';
-                if (strpos($suggestedName, 'array') !== false || substr($suggestedName, -2) === '[]') {
+                if (preg_match('/^(.*)<.*>$/', $suggestedName, $matches)) {
```
Comment (jrfnl) [inline src/Standards/Squiz/Sniffs/Commenting/FunctionCommentSniff.php:422]: 1. For performance optimization, I'd recommend having this condition as the _last_ condition in the if/elseif chain, not the first.
2. I'd also suggest optimizing the regex a little more, like so:
    ```suggestion
                    if (preg_match('/^([^<]+)<[^>]+>$/', $suggestedName, $matches)) {
    ```

File: src/Standards/Squiz/Tests/Commenting/FunctionCommentUnitTest.inc [hunk 2]
```diff
@@ -1042,7 +1042,19 @@ public function ignored() {
...
+ * Using generic as a type hint should satisfy a specified object parameter type.
+ * @see https://phpstan.org/blog/generics-in-php-using-phpdocs
+ *
+ * @param Collection<int, string> $values An object with int, string pairs.
+ *
+ * @return void
+ */
+public function specifiedArray1(Collection $values) {
+
+}// end specifiedArray1()
```
Comment (jrfnl) [inline src/Standards/Squiz/Tests/Commenting/FunctionCommentUnitTest.inc:1060]: Missing new line at end of file (also for fixed).

**Review by jrfnl — COMMENTED** [review]
Looks good, just two small notes.

**Comment by tm1000** [comment]
@jrfnl Did as you asked however I could not make the if/then last. I was able to make it second to last. The reasoning is because the last if/then is a "if not in list then do this". The suggestedName before the expression will never be in that array.

**Review by jrfnl — APPROVED** [review]

**Comment by jrfnl** [comment]
Linking issue #3589 for visibility.

**Comment by tm1000** [comment]
@jrfnl @gsherwood Now that its been about 9 months what is the process to get this merged?

**Comment by jrfnl** [comment]
@tm1000 Probably not what you want to hear, but patiently wait until Greg has some time to review this. I can review (which I did), but I don't have commit, so there's nothing more I can do.

**Review by bsandstrom — APPROVED** [review]

**Comment by WillemHoman** [comment]
@jrfnl would love to see this merged, is that going to be possible?

**Comment by tm1000** [comment]
Ref: https://github.com/PHPCSStandards/PHP_CodeSniffer/pull/750

**Pull request closed without merging**